### PR TITLE
[IOTDB-5174] Use filename format such as NodeID-Index rather than Endpoint-Index to track follower sync progress

### PR DESCRIPTION
**Pull request by OneSizeFitsQuorum**
see [jira5174](https://issues.apache.org/jira/browse/IOTDB-5174)

### CODE REVIEW

File: consensus/src/main/java/org/apache/iotdb/consensus/iot/logdispatcher/IndexController.java [hunk 1]
```diff
@@ -44,16 +48,23 @@ public class IndexController {
...
   private final long initialIndex;
 
   private final long checkpointGap;
 
-  public IndexController(String storageDir, String prefix, long initialIndex, long checkpointGap) {
+  public IndexController(String storageDir, Peer peer, long initialIndex, long checkpointGap) {
     this.storageDir = storageDir;
-    this.prefix = prefix + '-';
+    this.peer = peer;
+    this.prefix = peer.getNodeId() + "-";
```
Comment (xingtanzjr) [inline consensus/src/main/java/org/apache/iotdb/consensus/iot/logdispatcher/IndexController.java:61]: Suggest to use a static string variable to represent the spliter `"-"`

**Review by SpriCoder — APPROVED** [review]
LGTM~

File: consensus/src/test/java/org/apache/iotdb/consensus/iot/logdispatcher/IndexControllerTest.java [hunk 2]
```diff
@@ -58,28 +65,42 @@ public void testIncrementIntervalAfterRestart() {
...
     Assert.assertEquals(CHECK_POINT_GAP, controller.getCurrentIndex());
     Assert.assertEquals(CHECK_POINT_GAP, controller.getLastFlushedIndex());
 
     controller.updateAndGet(CHECK_POINT_GAP * 2 + 1);
     Assert.assertEquals(CHECK_POINT_GAP * 2 + 1, controller.getCurrentIndex());
     Assert.assertEquals(CHECK_POINT_GAP * 2, controller.getLastFlushedIndex());
   }
+
+  @Test
+  public void testUpgrade() throws IOException {
```
Comment (xingtanzjr) [inline consensus/src/test/java/org/apache/iotdb/consensus/iot/logdispatcher/IndexControllerTest.java:94]: Suggest to add another UT to process the upgrade where the peer's endpoint contains `-`.

File: consensus/src/main/java/org/apache/iotdb/consensus/iot/logdispatcher/IndexController.java [hunk 3]
```diff
@@ -124,6 +135,29 @@ private void persist() {
...
+  private void upgrade() {
+    File directory = new File(storageDir);
+    String oldPrefix = Utils.fromTEndPointToString(peer.getEndpoint()) + "-";
+    Optional.ofNullable(directory.listFiles((dir, name) -> name.startsWith(oldPrefix)))
+        .ifPresent(
+            files ->
+                Arrays.stream(files)
+                    .forEach(
+                        oldFile -> {
+                          long fileVersion = Long.parseLong(oldFile.getName().split("-")[1]);
```
[the remark this answers is not in the document]
Comment (OneSizeFitsQuorum) [reply consensus/src/main/java/org/apache/iotdb/consensus/iot/logdispatcher/IndexController.java:147]: In IoTDB 1.0, IoTConsensus cannot currently be restarted if hostname has -. So my assumption here is that if a real 1.0 user needs to upgrade, hostname will definitely not include `-`.

File: consensus/src/test/java/org/apache/iotdb/consensus/iot/logdispatcher/IndexControllerTest.java (hunk 2 above)
Comment (OneSizeFitsQuorum) [reply consensus/src/test/java/org/apache/iotdb/consensus/iot/logdispatcher/IndexControllerTest.java:94]: I have change the default Peer ip from `127.0.0.1` to `datanode-1.datanode-svc`

File: consensus/src/main/java/org/apache/iotdb/consensus/iot/logdispatcher/IndexController.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (OneSizeFitsQuorum) [reply consensus/src/main/java/org/apache/iotdb/consensus/iot/logdispatcher/IndexController.java]: Fixed

File: consensus/src/main/java/org/apache/iotdb/consensus/iot/logdispatcher/IndexController.java [hunk 4]
```diff
@@ -30,12 +32,16 @@
...
 import java.nio.file.Files;
+import java.util.Arrays;
+import java.util.Optional;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
 
 /** An index controller class to balance the performance degradation of frequent disk I/O. */
 @ThreadSafe
 public class IndexController {
 
+  public static final String separator = "-";
```
Comment (xingtanzjr) [inline consensus/src/main/java/org/apache/iotdb/consensus/iot/logdispatcher/IndexController.java:43]: uppercase ？

**Review by xingtanzjr — APPROVED** [review]

File: consensus/src/main/java/org/apache/iotdb/consensus/iot/logdispatcher/IndexController.java (hunk 4 above)
Comment (OneSizeFitsQuorum) [reply consensus/src/main/java/org/apache/iotdb/consensus/iot/logdispatcher/IndexController.java]: I just follow the `File.separator` style. Maybe this is trivial?
image

**Pull request merged**